src/app/icd/ICDClientInfoPersistentStorage.h

```diff
@@ -0,0 +1,165 @@
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/StorageKeyName.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * Delegate for providing storage key names and maximum key counter.
+ */
+class ICDStorageKeyDelegate
+{
+public:
+    /**
+     * Gets the storage key name for the specified index.
+     *
+     * @param aIndex The index of the key to retrieve.
+     *
+     * @return The storage key name.
+     */
+    virtual StorageKeyName GetKey(uint16_t aIndex) = 0;
+
+    /**
+     * @brief Gets the maximum key counter.
+     *
+     * @return The maximum key counter.
+     */
+    virtual uint16_t MaxKeyCounter() = 0;
+
+    virtual ~ICDStorageKeyDelegate(){};
+};
+
+/**
+ * The ICDClientInfoPersistentStorage class is an abstract interface that defines the operations
+ * for storing and retrieving ICD client information in persistent storage.
+ */
+class ICDClientInfoPersistentStorage
+{
+public:
+    /**
+     * Struct to store information about ICD Client Info.
+     */
+    struct ICDClientInfo
+    {
+        NodeId mPeerNodeId                 = kUndefinedNodeId;
+        uint32_t mStartICDCounter          = 0;
+        uint32_t mOffset                   = 0;
+        uint64_t mMonitoredSubject         = static_cast<uint64_t>(0);
+        Crypto::Aes128KeyHandle mSharedKey = Crypto::Aes128KeyHandle();
+
+        bool IsValid() { return mPeerNodeId != kUndefinedNodeId; }
+    };
+
+    /**
+     * Struct per identity representing persistent storage for ICD client information and key store
+     */
+    struct Storage
+    {
+        FabricIndex mFabricIndex                      = kUndefinedFabricIndex;
+        PersistentStorageDelegate * mpClientInfoStore = nullptr;
+        Crypto::SessionKeystore * mpKeyStore          = nullptr;
+        void Init(FabricIndex aFabricIndex, PersistentStorageDelegate * apClientInfoStore, Crypto::SessionKeystore * apKeyStore)
+        {
+            mFabricIndex      = aFabricIndex;
+            mpClientInfoStore = apClientInfoStore;
+            mpKeyStore        = apKeyStore;
+        }
+        Storage() {}
+        ~Storage() {}
+        Storage(Storage && aOther)
+        {
+            mFabricIndex             = aOther.mFabricIndex;
+            mpClientInfoStore        = aOther.mpClientInfoStore;
+            mpKeyStore               = aOther.mpKeyStore;
+            aOther.mpClientInfoStore = nullptr;
+            aOther.mpKeyStore        = nullptr;
+            aOther.mFabricIndex      = kUndefinedFabricIndex;
+        }
+        Storage & operator=(const Storage & aOther) = default;
+        bool IsValid() { return mFabricIndex != kUndefinedFabricIndex && mpClientInfoStore != nullptr && mpKeyStore != nullptr; }
+    };
+
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientInfoPersistentStorage(){};
+
+    /**
+     * Iterate through persisted ICD Client Info
```

Comment:
We would have two high-level delegate, one is from @erjiaqing  song's PR, which has ICDRegistrationDelegate, another is CheckInDelegate where it provides OnValidateCheckInMessage(System::PacketBufferHandle && payload) and OnReadyForSubscription or something similar. Both Delegate's implementation takes this ICDClientInfoPersistentStorageDelegate. If application don't wanna use ICDClientInfoPersistentStorageDelegate, and wanna implement the high-level API callbacks in CheckInDelegate and ICDRegistrationDelegate, it is fine. 
Here, it seems it is not wise to put the decryption logic in ICDClientInfoPersistentStorageDelegateImpl(Take checkin message buffer, and return the clientInfo..)...

Reply:
@bzbarsky-apple @erjiaqing

Reply:
I guess you wanna hide this iterator to impl "using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;" Actually  https://github.com/project-chip/connectedhomeip/blob/master/src/lib/support/CommonIterator.h#L34 provide pure virtual methods. So this delegate just enforce application to use abstract iterator, I am trying to limit the minimum implementation here... For clientInfos from multiple storages, we really need a iterator to traverse all keys and their metadata, but the methods for the iterator can be customized by application

Reply:
Having an abstract iterator is fine.  That part I am not concerned about.

Forcing use of "SessionKeystore" is not so fine: it's an insecure API that requires the raw key material to be handed around in cleartext, no?  And since we are not using either DeriveKey or DeriveSessionKeys, what, exactly, does SessionKeystore buy us?  In particular:

> it seems it is not wise to put the decryption logic in ICDClientInfoPersistentStorageDelegateImpl

I thinks we need to have it be possible for the decryption logic to work with keys that are in secure storage and that never appear in RAM in cleartext.  The APIs being proposed here we have now don't let us do that, right?

I do agree there is value in not having "must do it that way per spec" decryption code implemented by API consumers.  That means one of:

1. Structuring APIs so that the SDK can do the decryption but is completely agnostic as to the storage of the keys (i.e. it never generates them or creates key handles; it just asks the API consumer to do that and then uses the key handles).
2. Exposing helper functions as public API so the consumer can do the decryption correctly.

Right?

Similarly, forcing the use of persistent storage delegate, a low-level API that may not be a good fit for this situation, is not fine.  Again, what we are really doing is "we have a key and some associated metadata".  We should be letting people store that metadata however they want...

Reply:
Now I am just using alias SymmetricKeyStore from SessionKeystore...when new SymmetricKeyStore comes out, we would replace it.

Agree. just create new API, bool ValidateKey(System::PacketBufferHandle && aPayload, ICDClientInfo & aClientInfo), please see latest commit.

Reply:
@bzbarsky-apple, is that good for you? thanks

Reply:
1. Add virtual bool ValidateKey(const ByteSpan & aNonce, const ByteSpan & aEncryptedData, const ByteSpan & aMIC, ICDClientInfo & aClientInfo), which can help hide decryption logic in secure storage.
2. Remove CreateICDInfoEntry, Add SetKey and StoreICDInfoEntry, SetKey is only called when handling ICD registration during commissioning. StoreICDInfoEntry would be used to store/update icdInfo during commisioning and during handling check-in message.
3. Move ICDStorage related things into DefaultICDClientInfoPersistentStorage, instead of its delegate.